contracts/Comet.sol

```diff
@@ -1190,10 +1202,11 @@ contract Comet is CometMainInterface {
 
         uint collateralAmount = quoteCollateral(asset, baseAmount);
         if (collateralAmount < minAmount) revert TooMuchSlippage();
+        if (collateralAmount > getCollateralReserves(asset)) revert InsufficientReserves();
 
         // Note: Pre-transfer hook can re-enter buyCollateral with a stale collateral ERC20 balance.
         //       This is a problem if quoteCollateral derives its discount from the collateral ERC20 balance.
-        withdrawCollateral(address(this), recipient, asset, safe128(collateralAmount));
+        doTransferOut(asset, recipient, safe128(collateralAmount));
```

Comment:
I think this change actually makes the protocol even more susceptible to a pre-transfer hook attack. A malicious token can bypass the `if (collateralAmount > getCollateralReserves(asset)) revert InsufficientReserves();` to buy more than the collateral reserves with re-entrancy.

This is because the accounting check now happens separately and before the `doTransferOut`, while the old code did the accounting check and `doTransferOut` together in `withdrawCollateral`.

Reply:
Malicious token can buy more of itself than reserves? Interesting, we should think this through

Reply:
Discussed offline. Thankfully, only the `msg.sender` (`Comet`) can register a pre-transfer hook for a ERC777 token. BUT, this is still vulnerable to a malicious token, so we should probably update the documentation.

Reply:
We think this isn't susceptible issues from ERC777 since only sender can register pre-transfer hooks in that standard. For some questionable asset in the future which allows arbitrary pre-transfer behavior, it should not be listed, but at least the risk is isolated to the questionable asset